### ISSUE TITLE
High CPU (possibly stick-tables related) on 2.9.6

### ISSUE THREAD

**Issue by felipewd** [haproxy/haproxy#2475]
### Detailed Description of the Problem

Hello,

We've deployed 2.9.6 on production to confirm the stick-tables locking speedup that Willy provided on our old ticket #2164 to confirm the findings there. The TL;DR on that original issue was a stick-table with `2s` expire showed a lot of contention and the v2.9 branch got a little love to improve it.

Anyway, we've deployed it and reduced the `expire 5m` to `expire 3s` on our stick-tables used for JWT validation and source IP tracking, which for our application provides the best functionality (we're trying to track if some user is using the same JWT with 2+ different IPs).

We still found some noticeable jumps on %user, with threads going to %93 CPU usage on htop.

A few details for `show activity`:

```
thread_id: 43 (1..64)
date_now: 1709413203.758988
uptime_now: 20938.381639
ctxsw: 1059900712 [ 152142307 150625894 151900590 148514024 150405929 150381069 152168828 155564366 153530081 155718180 150791456 154923106 146063080 149897691 149532604 148006493 148874456 151399921 151222301 153246656 154387195 154070032 151335583 154380463 149615486 154537309 149784202 149081318 150148445 152042381 152684069 154391098 146430560 150246193 146912229 142069967 156153731 146732894 151052987 152456022 148304368 146662993 153277332 155161412 149727080 153445931 153515168 150165709 150702567 151952145 146555681 153452766 147680830 150865417 150081076 149776359 150960907 148678741 145705432 149128197 149659553 148582606 148866869 153504969 ]
tasksw: 142273063 [ 2040930 2006722 2102194 2162909 2043877 2037233 2033707 2019324 2134746 2080387 2039436 2031499 2155295 2104731 2125615 2254622 2068937 2092287 1990434 1941008 2052983 2056923 2146436 2128069 2111554 2093287 2089999 2005316 2026765 2042748 2117492 2088045 2513155 2507462 2493446 2510348 2577777 2390311 2512170 2553359 2064367 2132576 2143065 2044857 2065034 2098769 2116046 2041354 2161228 2093335 2143043 2054834 2081640 2053676 2034246 2017683 2792100 2943752 2920230 2784583 2725205 2759360 2694884 2823658 ]
empty_rq: 60971177 [ 963120 469553 884252 828394 767211 693402 575760 855013 884876 869200 863236 528249 813401 654240 423360 750395 845084 760559 575484 839398 907272 873924 466317 668386 900595 844877 506739 412428 884126 619070 804452 739364 1536034 1200937 1067966 1419159 1463138 1580341 1327845 1555589 1021390 846781 919361 1033144 635603 985401 740824 1064348 758564 680398 814289 908896 857034 876639 566252 458355 1774932 1743666 1585989 1369036 1465744 1835399 1683701 1722685 ]
long_rq: 92031546 [ 1479906 1410312 1414774 1380311 1377988 1412075 1419557 1418509 1499552 1492577 1455530 1512485 1381036 1406079 1450453 1349475 1328600 1387056 1366311 1423935 1418415 1455038 1411483 1457242 1434607 1520651 1497648 1349174 1413711 1463898 1472984 1447890 1432566 1530656 1422686 1374781 1601748 1482373 1477722 1518879 1396414 1429210 1402522 1524354 1406196 1437254 1490583 1420574 1443241 1309886 1372368 1514296 1359761 1414064 1399082 1380214 1427602 1436674 1379285 1446774 1533131 1536261 1486524 1534603 ]
curr_rq: 3126 [ 65 88 82 0 68 0 0 102 70 90 40 5 83 94 0 50 0 0 0 109 95 92 0 0 25 0 95 90 0 112 101 87 96 76 2 81 95 2 72 34 0 119 1 84 0 65 27 86 54 1 0 65 91 96 0 89 91 0 95 0 29 32 0 0 ]
loops: 1251758922 [ 19700203 19239727 19607717 19601831 19711699 19631631 20041375 19866252 19696113 20039741 18978684 19558522 19376730 20211561 18421685 19919372 19944303 19329353 20004442 19936658 20470933 19512455 19220036 19651022 19556332 19461674 18694627 18870033 19330417 19082246 18984340 19547688 19707726 19198364 19209886 18963601 20911063 19594240 19436258 19901832 19627677 18429643 20757261 20514797 19565461 20144554 19418195 19423059 19000315 20174282 19390067 19064312 20043166 19399299 19459133 18967345 20066989 20106543 19585657 19321939 18733588 19650841 18791289 20001138 ]
wake_tasks: 92144441 [ 1481550 1412012 1416578 1382156 1379763 1413703 1421246 1420445 1501289 1494362 1457269 1514208 1382861 1407757 1452320 1351309 1330383 1389048 1368021 1425649 1420175 1456916 1413280 1459104 1436349 1522565 1499342 1350826 1415448 1465651 1474747 1449664 1434327 1532423 1424396 1376580 1603438 1484005 1479531 1520658 1398064 1431005 1404167 1526055 1407863 1439065 1492304 1422196 1445132 1311677 1374218 1516115 1361468 1415967 1400780 1381936 1429467 1438462 1381004 1448564 1534966 1537901 1488253 1536458 ]
wake_signal: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
poll_io: 1187689650 [ 18932954 18514535 18876539 18894058 19007262 18887412 19303180 19150656 18919220 19272347 18222902 18766509 18666070 19480344 17673534 19228894 19279302 18640941 19314575 19195857 19743518 18771229 18485165 18906327 18794908 18678148 17887728 18188655 18608696 18328220 18233709 18810710 18154827 17642807 17761435 17475300 19272710 17993353 17916144 18289969 18717214 17511081 19875156 19530258 18661769 19222471 18466961 18509922 18262501 19512937 18690082 18280157 19334285 18681766 18723252 18268611 18209607 18232724 17790397 17504826 16861129 17680747 16911079 18080069 ]
poll_exp: 355467 [ 3 2 3 2 1 4 3 2 2 2 0 1 3 2 1 2 3 2 4 3 8 1 1 0 1 7 4 0 3 3 0 1 16669 15973 16681 15681 18821 15105 16494 17248 1825 2187 1525 1595 1999 1203 1008 1592 3 3 0 4 1 2 2 3 24776 24408 24776 33121 26968 23127 25087 27506 ]
poll_drop_fd: 541137 [ 7107 8436 8330 8203 10512 7535 7600 9516 7654 8337 7483 6505 10018 11118 8267 8483 7525 8532 6697 7388 10933 6968 9341 8822 8247 7080 9238 7085 7406 8741 6805 7227 9834 8064 8749 10212 7255 7902 9076 10068 7996 9002 8119 7622 7638 8330 7013 7027 10162 7172 9692 6551 10455 7800 9799 8766 8161 11754 10260 9931 9048 9149 7377 8014 ]
poll_skip_fd: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
conn_dead: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
stream_calls: 92583788 [ 1425633 1403154 1468679 1508223 1430272 1416187 1421386 1415924 1483257 1459787 1421959 1425515 1509478 1466156 1496221 1558409 1448339 1472588 1394658 1355749 1427371 1440004 1490199 1490854 1475565 1466353 1459338 1409792 1411275 1423050 1476990 1456280 1442104 1464108 1442617 1473763 1437632 1397715 1458525 1477919 1402978 1469937 1445488 1390167 1411434 1428055 1439109 1379981 1506816 1453484 1507214 1436509 1443638 1450149 1433315 1415458 1393718 1506133 1515615 1444262 1439463 1424320 1421844 1421673 ]
pool_fail: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
buf_wait: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
cpust_ms_tot: 20110014 [ 325414 313068 314658 313017 311370 313013 314768 314739 316689 314967 320575 317388 314996 316495 312737 309580 315902 316636 313309 318315 317401 312693 309719 315874 306892 313353 314743 314494 312228 314234 318482 312404 307708 314825 313984 316207 313004 313719 312444 313506 313528 316700 315375 318561 309710 318427 314805 312469 319714 309565 309530 311990 316347 309417 311655 312425 316498 308766 316080 313655 314523 317693 312339 314692 ]
cpust_ms_1s: 4250 [ 58 61 57 66 68 66 75 88 92 86 80 52 61 63 82 53 69 54 59 93 60 68 69 35 64 53 84 62 69 81 57 71 66 84 53 72 84 84 75 69 75 66 82 63 30 63 86 28 74 40 82 53 61 34 45 46 91 83 61 75 91 77 20 81 ]
cpust_ms_15s: 48456 [ 734 785 671 850 681 787 728 770 781 725 792 735 816 874 775 798 878 814 822 870 711 811 663 766 677 709 765 644 830 870 797 766 683 747 636 768 768 846 838 698 811 737 810 763 748 835 779 798 735 617 666 787 776 670 595 687 799 675 768 682 764 781 773 721 ]
avg_cpu_pct: 57 [ 54 51 60 38 54 30 69 52 37 89 73 53 52 64 45 67 27 41 25 51 69 86 78 46 84 35 63 88 61 61 53 47 76 45 61 90 62 52 41 48 66 64 64 80 27 45 87 51 63 22 24 43 56 23 82 54 35 67 90 76 57 73 54 71 ]
avg_loop_us: 870 [ 1083 903 1214 701 749 392 812 734 762 995 887 592 746 1515 506 1066 473 437 483 1118 999 1035 410 800 1089 645 707 979 644 714 1172 620 1406 719 952 1713 750 798 783 779 1384 917 460 593 470 1033 1610 617 1396 608 248 694 1099 1040 954 510 807 831 940 906 1282 1137 831 1426 ]
accepted: 10299085 [ 427074 41862 302311 128618 81097 62374 46795 198067 452592 298418 200461 46710 116111 57016 39148 78146 196011 76595 52651 129174 451318 287779 35799 52018 480066 185325 38852 40649 301826 53155 110691 78786 191984 49012 39711 76990 129143 450709 59255 289092 191913 53139 78405 312400 38511 123115 46000 444698 77034 61279 116320 452142 198902 300639 43742 37124 307135 186367 59198 45185 51183 429924 83451 127888 ]
accq_pushed: 10299085 [ 164087 155940 165154 166490 160879 159641 158612 156171 163514 164393 156339 152519 164438 171872 158763 156308 160200 160054 156529 155662 162705 161417 167306 158162 159366 159370 165421 155245 159011 165043 159524 164406 168771 160409 163284 158934 158170 154618 161991 161811 157884 159851 164833 154780 162286 162614 161517 164667 168514 162113 166742 157379 162201 153376 157400 160666 157722 169882 166163 163689 154387 159063 157500 161327 ]
accq_full: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
accq_ring: 1 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
fd_takeover: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
check_adopted: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
check_started: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
check_active: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
check_running: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
```

And a 70s profiling:

```
Tasks activity over 71.795 sec till 0.000 sec ago:
  function                      calls   cpu_tot   cpu_avg   lat_tot   lat_avg
  sc_conn_io_cb              42516963   12.33m    17.40us   1.403d    2.851ms <- sc_app_chk_rcv_conn@src/stconn.c:787 tasklet_wakeup
  h1_io_cb                    4436164   44.55s    10.04us   1.369h    1.111ms <- sock_conn_iocb@src/sock.c:877 tasklet_wakeup
  sc_conn_io_cb               4232354   3.447m    48.87us   8.740h    7.434ms <- h1_wake_stream_for_recv@src/mux_h1.c:3193 tasklet_wakeup
  process_stream               208912   24.38s    116.7us   11.36m    3.263ms <- sc_notify@src/stconn.c:1139 task_wakeup
  h1_io_cb                     151970   2.067s    13.60us   1.997m    788.5us <- conn_subscribe@src/connection.c:832 tasklet_wakeup
  process_stream               132123   24.49m    11.12ms   8.339m    3.787ms <- stream_new@src/stream.c:578 task_wakeup
  h1_timeout_task               76072   54.86ms   721.0ns   2.089m    1.648ms <- h1_release@src/mux_h1.c:1060 task_wakeup
  h1_io_cb                      66029   6.074s    91.98us   43.43s    657.7us <- sock_conn_iocb@src/sock.c:856 tasklet_wakeup
  process_stream                63181   916.6ms   14.51us   35.14s    556.2us <- run_tasks_from_lists@src/task.c:657 task_queue
  sc_conn_io_cb                 52667   786.5ms   14.93us   54.44s    1.034ms <- h1_wake_stream_for_send@src/mux_h1.c:3203 tasklet_wakeup
  accept_queue_process          34508   656.7ms   19.03us   5.568m    9.681ms <- listener_accept@src/listener.c:1476 tasklet_wakeup
  process_peer_sync             26262   19.79ms   753.0ns   43.32s    1.649ms <- stktable_touch_with_exp@src/stick_table.c:529 task_wakeup
  process_table_expire          20034   2.123m    6.359ms   1.703s    84.99us <- run_tasks_from_lists@src/task.c:657 task_queue
  process_stream                 2693   22.83ms   8.478us   1.445s    536.7us <- sc_notify@src/stconn.c:1154 task_queue
  h1_timeout_task                 221   12.27ms   55.52us   334.7ms   1.514ms <- h1_init@src/mux_h1.c:993 task_queue
  h1_io_cb                        183   748.2us   4.088us   425.3ms   2.324ms <- h1_shutw_conn@src/mux_h1.c:4201 tasklet_wakeup
  process_stream                  158   1.245ms   7.879us   145.4ms   920.2us <- sc_app_chk_snd_conn@src/stconn.c:858 task_wakeup
  sc_conn_io_cb                   106   2.854ms   26.93us   321.4ms   3.032ms <- sock_conn_iocb@src/sock.c:877 tasklet_wakeup
  accept_queue_process             86   239.2us   2.781us   848.9ms   9.870ms <- accept_queue_process@src/listener.c:175 tasklet_wakeup
  task_run_applet                  77   1.332s    17.30ms   12.51ms   162.4us <- sc_applet_create@src/stconn.c:512 appctx_wakeup
  task_run_applet                  73   134.9us   1.847us   216.3ms   2.963ms <- sc_app_shut_applet@src/stconn.c:918 appctx_wakeup
  h1_timeout_task                  53   2.177ms   41.08us   157.1ms   2.964ms <- h1_refresh_timeout@src/mux_h1.c:632 task_queue
  task_run_applet                  37   142.9ms   3.862ms   87.31us   2.359us <- sc_app_chk_snd_applet@src/stconn.c:977 appctx_wakeup
  ssl_sock_io_cb                   14   5.568ms   397.7us   796.3us   56.88us <- sock_conn_iocb@src/sock.c:877 tasklet_wakeup
  h1_io_cb                          9   1.832ms   203.6us   1.750ms   194.5us <- ssl_sock_io_cb@src/ssl_sock.c:6637 tasklet_wakeup
  sc_conn_io_cb                     7   3.008us   429.0ns   27.29ms   3.899ms <- sock_conn_iocb@src/sock.c:856 tasklet_wakeup
  ssl_sock_io_cb                    7   3.391us   484.0ns   3.373ms   481.9us <- conn_subscribe@src/connection.c:832 tasklet_wakeup
  ssl_sock_io_cb                    7   2.318ms   331.2us   2.385ms   340.7us <- ssl_sock_start@src/ssl_sock.c:5888 tasklet_wakeup
  task_run_applet                   4   185.5ms   46.38ms   272.1us   68.03us <- sc_app_chk_rcv_applet@src/stconn.c:957 appctx_wakeup
  ```

Also @rnsanchez collected a flamegraph which shows stick-table unlocks taking up 50% of cycles.
![haproxy-2 9 6-9eafce5-20240302-2102 (1)](https://github.com/haproxy/haproxy/assets/547142/7d493c70-5b90-40b7-a5e9-5daf85d5094f)

### Expected Behavior

less %user on threads

### Steps to Reproduce the Behavior

At peak, this machine got ~85Gbps of production traffic, with a JWT validation + tracking on each request.,

### Do you have any idea what may have caused this?

stick-table locking, probably

### Do you have an idea how to solve the issue?

We wish :-) 

### What is your configuration?

```haproxy
global
    nbthread 64
    hard-stop-after 7d
    maxconn 900000
    ulimit-n 3150919
    user haproxy
    group haproxy
    daemon
    unix-bind user haproxy mode 777


    node haproxy-frontend

    localpeer peer_reload 

    tune.idletimer 1000
    tune.bufsize 104857
    tune.pipesize 262144

    stats socket /var/run/haproxy_conexoes.sock mode 666 level admin expose-fd listeners
    stats timeout 10m

    log /var/tmp/logtape.unix local0

    h1-case-adjust content-length Content-Length
    h1-case-adjust content-type Content-Type
    h1-case-adjust content-range Content-Range
    h1-case-adjust cache-control Cache-Control
    h1-case-adjust server Server
    h1-case-adjust date Date

...
backend client_stats
stick-table type string len 512 size 2m store http_req_rate(2s),conn_cur expire 10m peers peer_local

backend hyper_streaming_jwt_client
    timeout server 19s

acl tem_token url_param(sjwt)  -m found
acl tem_ts var(txn.path) -m sub .ts
acl tem_m3u8 var(txn.path) -m sub .m3u8
acl tem_mp4 var(txn.path) -m sub .mp4

http-request deny status 200 content-type video/MP2T file /home/felipewd/client.ts  if tem_token !{ var(txn.bearer),jwt_verify("HS256","PdSgVkYp3s6v9yBEHMcQfThWmZq4t7wzCFJaNdRgUkXn2r5u8xADG") 1 }
...

acl whitelist_usuario var(req.usuario) -m str "test"

stick-table type string len 512 size 2m store http_req_rate(2s),conn_cur expire 3s peers peer_local
http-request track-sc0 var(req.usuario) if !tem_ts !whitelist_usuario
http-request track-sc0 var(req.usuario) if  tem_ts tem_token !whitelist_usuario
http-request track-sc1 var(req.usuario_ip) if  !tem_ts !whitelist_usuario
http-request track-sc1 var(req.usuario_ip) if  tem_ts tem_token !whitelist_usuario
http-request track-sc2 var(req.usuario) table client_stats if tem_token

acl alguma_conexao_mpegts sc0_conn_cur gt 1
acl negar_conexao_por_ip_mpegts var(req.usuario_ip),table_conn_cur lt 2

acl alguma_req_http sc0_http_req_rate gt 1
acl negar_conexao_por_req var(req.usuario_ip),table_http_req_rate lt 2

acl limite_conexoes_http sc0_http_req_rate gt 80
acl limite_conexoes_jwt sc0_conn_cur gt 50

http-request deny status 200 content-type video/MP2T file /home/felipewd/client.ts  if !tem_m3u8 limite_conexoes_jwt || !tem_m3u8 limite_conexoes_http || !tem_m3u8  alguma_conexao_mpegts negar_conexao_por_ip_mpegts || !tem_m3u8 alguma_req_http negar_conexao_por_req 


    server v01 abns@/tmp/test.sock
```

### Output of `haproxy -vv`

```plain
haproxy -vv:

HAProxy version 2.9.6-9eafce5 2024/02/26 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.9.6.html
Running on: Linux 6.0.6 #4 SMP PREEMPT_DYNAMIC Wed Nov 2 17:36:28 -03 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_NS=1 USE_PCRE=1 USE_PCRE_JIT=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_AWSLC -OPENSSL_WOLFSSL -OT +PCRE -PCRE2 -PCRE2_JIT +PCRE_JIT +POLL +PRCTL -PROCCTL -PROMEX -PTHREAD_EMULATION -QUIC -QUIC_OPENSSL_COMPAT +RT +SHM_OPEN -SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL +ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=128).
Built with OpenSSL version : OpenSSL 1.1.1q  5 Jul 2022
Running on OpenSSL version : OpenSSL 1.1.1q  5 Jul 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : SSLv3 TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.4.6
Built with network namespace support.
Built with zlib version : 1.2.13
Running on zlib version : 1.2.13
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.45 2021-06-15
Running on PCRE version : 8.45 2021-06-15
PCRE library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 12.2.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : none

Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

uname -a:

```
Linux ndt-spo-07 6.0.6 #4 SMP PREEMPT_DYNAMIC Wed Nov 2 17:36:28 -03 2022 x86_64 Intel(R) Xeon(R) Gold 6338N CPU @ 2.20GHz GenuineIntel GNU/Linux
```
```


### Last Outputs and Backtraces

```plain
.
```

**Comment by rnsanchez** [comment]
Here's `perf annotate`:
![image](https://github.com/haproxy/haproxy/assets/87608/6f74098f-f602-4b56-8555-d884ee915750)

**Comment by wtarreau** [comment]
Hi Felipe, Ricardo!

Thanks for this nice feedback. Here it's definitely the read lock on the stick-table that's taking a lot of time because most of the time another thread is already keeping the table write-locked for purging an entry or just inserting one. Given that stktable_get_entry() appears on two lock branches here, I suspect that what happens is that the vast majority of waiters are in stktable_lookup_key() which just takes a read lock, and that a few are form __stktable_store() that's directly inlined in stktable_get_entry(), and which is a write lock. The creation of a few entries is making other threads wait.

I'm wondering if we couldn't figure a way to batch removal of purged entries: we would allow them to linger there for a longer time than configured hoping they get a chance to be recycled and reused, otherwise we'd have to batch-remove them, taking a lock only once for a few to a few tens of entries. Note that I'm saying this without looking in details, of course, but that might be possible.

Another possibility could be to "shard" the keys: instead of having a single tree head in the table, we could have, say, 64, and hash the key so that there are 64 independent sub-tables. Of course this would require to move the expirations there as well, so that we can have one lock per sub-table instead of one shared lock. Maybe that approach would be more scalable by the way, and should be considered with priority as it doesn't sound super complicated.

**Comment by felipewd** [comment]
Hi @wtarreau thanks for the quick reply!

The sharding idea sounds great! Call me crazy, but this "one lock per sub-table instead of 1 shared lock" just reminded me of the old BKL removal patches back in the day :-) 

Anyway, this approach does sound more scalable indeed. Let us know how we can help

**Comment by wtarreau** [comment]
We already do that at other places, but not all places are compatible with this. Stick-tables are special in that there is no relation between keys, so we can afford to store them at different places. In the worst case it will be the "show table" that will list them in whatever order. No big deal.

**Comment by felipewd** [comment]
Sounds great! And it sounds future-proof in relation to these bigger core machines.

We intend on experimenting with 200-300Gbps machines this year, and this approach sounds great for this kind of traffic.

**Comment by wtarreau** [comment]
Do you have a test platform to test totally experimental code ?  I gave it a try to see if there was any showstopper, and it went relatively smoothly. The only limit is that "show table" as well as the Lua table scan currently only list the first shard. But I'd be interested in knowing if:
  - you observe some crashes, indicating that some of the locks also protect other areas and cannot be sharded like this;
  - it gets significantly better or not (at least in callgraphs)

**Comment by felipewd** [comment]
@wtarreau we can put it in a smaller production machine to give it a go, yes :-)

**Comment by wtarreau** [comment]
OK, here it comes for 3.0-master, I think it should be OK on top of 2.9 since these parts do not change often. It has received very little testing (it builds and passes the regtests that are not sensitive to show table).
[0001-EXP-stktable-try-to-split-the-keys-across-multiple-s.patch.txt](https://github.com/haproxy/haproxy/files/14486082/0001-EXP-stktable-try-to-split-the-keys-across-multiple-s.patch.txt)

**Comment by felipewd** [comment]
hey @wtarreau it crashed with 100% cpu after just a few seconds:

a [crash.log](https://github.com/haproxy/haproxy/files/14486604/crash.log) from core attached.

We have the bin + core if you want.

**Comment by wtarreau** [comment]
Based on the dump, something's got corrupted, too many waiters at the same place. It's possible that I missed an unlock somewhere, I'll need to recheck in depth. Hmmm I also just found stksess_kill_if_expired() in stick-table.h that needs to be converted to use the shards as well. OK let's forget that patch for now, I'll have more work to do on it. Thanks for the test. Don't worry, I'm still having hopes, but at the moment I'm quite busy on a ton of other stuff that I absolutely need to finish.

**Comment by wtarreau** [comment]
OK better now, I fixed the other points I found. A test config on a 24-core EPYC jumped from 358k rps to 2.35M, which tells me it mostly works :-)

However, the peers code takes the table lock and will continue to access the entries unlocked. I don't know this part well enough to do a quick hack there. Thus if you want to test it, I suggest that you comment all your peers (or point them to closed ports so that they don't learn anything).

Here's the updated patch, in case you can give it a try. Sorry, it has the same name (the updated one mentions peers at the end of the commit message).

[0001-EXP-stktable-try-to-split-the-keys-across-multiple-s.patch.txt](https://github.com/haproxy/haproxy/files/14492353/0001-EXP-stktable-try-to-split-the-keys-across-multiple-s.patch.txt)

**Comment by felipewd** [comment]
hello @wtarreau still no luck with crashes right after start (with 100% CPU).

I've applied it on 2.9.6 and 3.0-dev4....got core dumps from both of them if you want to take a look.

Backtraces:
[crash-2.9.log](https://github.com/haproxy/haproxy/files/14497877/crash-2.9.log)
[crash-3.0-dev4.log](https://github.com/haproxy/haproxy/files/14497879/crash-3.0-dev4.log)

**Comment by felipewd** [comment]
Oops, sorry, my bad!

Like a moron, I went straight to the patch and didn't even read the "don't use peers" part. I've commented them out and it works fine, yes!

We'll report back after a few hours to collect some traffic.

**Comment by wtarreau** [comment]
Excellent, I'm indeed impatient to know how much it improves the situation for you!

**Comment by rnsanchez** [comment]
Hi Willy,

I've repeated our data collection for a random HAProxy thread (patched), and it is a totally different scenario.  It is worth noting that the server was operating at only 55 Gbps during data collection, but still, `pl_wait_unlock` is taking so little CPU that is barely noticeable.  This looks like solid improvement!  :-)

![haproxy-2 9 6-patch-20240305-2215](https://github.com/haproxy/haproxy/assets/87608/9a5f75b8-b163-4572-9290-200e2dbf8463)

Here's considering only HAProxy (no kernel/libc):

![image](https://github.com/haproxy/haproxy/assets/87608/c072f30c-c772-49a3-9cdd-8bff0330cdb9)

**Comment by wtarreau** [comment]
Super cool, thanks for the measurements! It confirms that we must absolutely finish that one for 3.0, it seems like a low-hanging fruit (unless there are trouble on the peers side of course, which I hope not). In the worst case it shows too difficult to adapt the peers code to it, I'm still having a fallback in mind: I think we could use read-locks on the table everywhere we touch any shard of the table, so that it remains exclusive with peers updates. But that would still be a loss so I'd prefer to make sure peers are properly handled.

**Comment by felipewd** [comment]
Hey @wtarreau, just reaching out to ask something about these patches: `show table <stick tables>` now shows significantly fewer entries than before. 

I'm 99% sure this is because it's only 1 shard, but just wanted to make sure. Because we use these dumps for some internal stats and the fewer entries have puzzled some people :-) 

Thanks!

**Comment by wtarreau** [comment]
Yes, @felipewd I thought I mentioned it but the patch doesn't include the iteration over each tree head in show table (same problem as for peers).

**Comment by wtarreau** [comment]
hi guys,

just so that you know, I've found a small bug in the patch I sent you that may occasionally crash when entries are expired. I'm also seeing that the update_lock in stktable_touch_with_exp takes 43% of the CPU (even just the read lock) and I suspect I could get rid of it entirely for the read part if peers mention they've seen the node. As an experiment I tried to just comment out the read part and the perf went x2 when using peers :-)  So I think we're on something there. I've also found the thing that's incorrect wrt the peers locking and I think I'm seeing how to address it. Stay tuned.

**Comment by felipewd** [comment]
hi @wtarreau  not sure if it's related, but we've had 2 crashes today under heavy (89Gbps continuous for ~30min) pressure.

We couldn't get a core to properly find out, since it only happened twice on different machines...but both had a lot of active sessions (~65,000) which would use our 3s key expire all the time.

We'll try and set something up to track it down, but if you have a patch to test this crash, we can maybe validate it (depending on the pressure, I suppose).

**Comment by wtarreau** [comment]
I hoped to be able to give you a final patch yesterday evening but I faced a strange bug in the dump, reason why I postponed. In the worst case I'll send you the updated patch without the dump part while I'm trying to figure the problem there. I doubled the request rate with peers :-)  In fact, before peers would force a read lock to be used quite a lot from stick tables, and now it's rarely used.

**Comment by wtarreau** [comment]
Hi guys,

here are some news. I could redo the patch and merge it in 3.0-dev (1a088da7c2). There are a few to take before it to be effective though, so the series should rather be:

- ed45d13321 BUG/MEDIUM: stick-table: use the update lock when reading tables from peers
- 6a2f09de1c OPTIM: peers: avoid the locking dance around peer_send_teach_process_msgs()
- 4c1480f13b MINOR: stick-tables: mark the seen stksess with a flag "seen"
- 864ac31174 OPTIM: stick-tables: check the stksess without taking the read lock
- 1a088da7c2 MAJOR: stktable: split the keys across multiple shards to reduce contention

The first one is already backported in the latest 2.9 BTW. The other ones are normally not planned for backports. I haven't checked if all of these backport well or not. I think they should given that we don't touch this area often, but if you encounter issues, just let me know. For the record, the last one multiplied the perf by 6.5 for me on 80 threads, and the previous one made peers almost costless on the most common path ;-) YMMV of course.

**Comment by felipewd** [comment]
hey @wtarreau we've deployed these patches on v2.9 and will monitor tonight under heavy traffic. But nice work! We're very happy to be a small part of such great improvements :-) 

Thanks!

**Comment by wtarreau** [comment]
Rest assured that as long as you test our patches, you're all but a small part! Live testing under the worst conditions is the most difficult thing to get on development and can condition the merging of certain features. So please be proud of your participation, it's really much appreciated, you've already reported quite a number of shortcoming affecting modern versions that helped us improve the overall code, that's cool. I hope you'll continue like this ;-)

**Comment by felipewd** [comment]
hey @wtarreau just a quick feedback: We couldn't get a ton of traffic (~65-67Gbps peak), but everything went smoothly. Even the previous 2 crashes didn't happen today. So we're good on all fronts :-) 

I'm closing this ticket now. Thanks a lot!

**Comment by wtarreau** [comment]
That's a great feedback, thanks. For those stumbling upon this issue in the future, here's a summary of the situation: 2.9 works as designed but under some high stress with many cores, there can be contention on the stick-tables locks (main one and update lock when using peers), that affects all versions up to 2.9. Apparently the bwlim filters are increasing the level of stress on the stick-tables thus the contention. These changes are not intended to be backported but 3.0 will be significantly better regarding this.

**Comment by wtarreau** [comment]
Guys, you may be interested in having a look at issue #2525. The patches to improve the stick-tables locking unveiled an older one that affects 2.9 with peers. It's difficult to reproduce so you may possibly not be affected. There's a small patch that avoids passive peers from staying at 100% CPU when receiving a constant flow of updates, and another one (that will not trivially backport, I'll have to adapt it) that fixes a remaining race between peers and stick-tables that causes single-byte memory corruption hence dirty crashes. We've had it since 2.9 but it seems that the locking overhead prevented the problem from appearing, and that it's no longer the case :-)  The fix will be backported into the next 2.9 anyway so that your patch series will remain valid.

**Comment by felipewd** [comment]
hey @wtarreau we just saw the commits on 3.0 branch and were discussing them.

They seem totally relevant to our setup and we've also decided to give 3.0 a try in production for early testing also. Anything we can do to help catch nasty situations like these as soon as possible.

We'll grab 3.0-head now and report back if we find anything :-)

Thanks again!

**Comment by wtarreau** [comment]
Awesome, that's even better (and simpler). The most important rule to keep in mind with -dev is to make it easy to roll back. Usually it works pretty well and is very smooth, but once or twice per major version there's a nasty regression (like this one) and in this case you just want to be able to quickly revert to a known good version without having to hack config files etc. But I have no doubt that this is probably a detail for you ;-)

**Comment by felipewd** [comment]
@wtarreau Well, that didn't take long. We crashed right away. :-) 

In a place we didn't expect, actually.

We can create a separate issue if you want, just posting it here for fast reporting:

```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007fb0cb1377d6 in epoll_wait () from /lib64/libc.so.6
[Current thread is 1 (Thread 0x7fb0caf67180 (LWP 1164953))]
(gdb) bt f 
#0  0x00007fb0cb1377d6 in epoll_wait () from /lib64/libc.so.6
No symbol table info available.
#1  0x0000000000433471 in _do_poll (p=<optimized out>, exp=<optimized out>, wake=<optimized out>) at src/ev_epoll.c:232
        timeout = 7
        status = <optimized out>
        fd = <optimized out>
        count = <optimized out>
        updt_idx = <optimized out>
        wait_time = 7
        old_fd = <optimized out>
#2  0x0000000000589b58 in run_poll_loop () at src/haproxy.c:3130
        next = <optimized out>
        wake = <optimized out>
        __func__ = <optimized out>
#3  0x000000000058a1f7 in run_thread_poll_loop (data=<optimized out>) at src/haproxy.c:3272
        ptaf = <optimized out>
        ptif = 0x71c6b0 <per_thread_init_list>
        ptdf = <optimized out>
        ptff = <optimized out>
        init_left = 0
        init_mutex = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 0, __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, 
          __align = 0}
        init_cond = {__data = {{__wseq = 1143, __wseq32 = {__low = 1143, __high = 0}}, {__g1_start = 1047, __g1_start32 = {__low = 1047, __high = 0}}, __g_refs = {0, 0}, __g_size = {0, 0}, 
            __g1_orig_size = 192, __wrefs = 0, __g_signals = {0, 0}}, __size = "w\004\000\000\000\000\000\000\027\004", '\000' <repeats 22 times>, "\300", '\000' <repeats 14 times>, __align = 1143}
#4  0x0000000000430f61 in main (argc=<optimized out>, argv=0x7ffe1d569948) at src/haproxy.c:3987
        err = <optimized out>
        retry = <optimized out>
        limit = {rlim_cur = 18446744073709551615, rlim_max = 18446744073709551615}
        pidfd = <optimized out>
        intovf = <optimized out>
        msg = <optimized out>
```

We're using 3.0 up to c0ee2d78d739

**Comment by rnsanchez** [comment]
Also, very odd to crash _inside_ `epoll_wait`:
```
(gdb) disassemble 
Dump of assembler code for function epoll_wait:
   0x00007f926fde4780 <+0>:     mov    %ecx,%r10d
   0x00007f926fde4783 <+3>:     mov    %fs:0x18,%eax
   0x00007f926fde478b <+11>:    test   %eax,%eax
   0x00007f926fde478d <+13>:    jne    0x7f926fde47a0 <epoll_wait+32>
   0x00007f926fde478f <+15>:    mov    $0xe8,%eax
   0x00007f926fde4794 <+20>:    syscall 
   0x00007f926fde4796 <+22>:    cmp    $0xfffffffffffff000,%rax
   0x00007f926fde479c <+28>:    ja     0x7f926fde47f8 <epoll_wait+120>
   0x00007f926fde479e <+30>:    ret    
   0x00007f926fde479f <+31>:    nop
   0x00007f926fde47a0 <+32>:    sub    $0x28,%rsp
   0x00007f926fde47a4 <+36>:    mov    %edx,0x18(%rsp)
   0x00007f926fde47a8 <+40>:    mov    %rsi,0x10(%rsp)
   0x00007f926fde47ad <+45>:    mov    %edi,0xc(%rsp)
   0x00007f926fde47b1 <+49>:    mov    %ecx,0x1c(%rsp)
   0x00007f926fde47b5 <+53>:    call   0x7f926fd60d50 <__libc_enable_asynccancel>
   0x00007f926fde47ba <+58>:    mov    0x1c(%rsp),%r10d
   0x00007f926fde47bf <+63>:    mov    0x18(%rsp),%edx
   0x00007f926fde47c3 <+67>:    mov    %eax,%r8d
   0x00007f926fde47c6 <+70>:    mov    0x10(%rsp),%rsi
   0x00007f926fde47cb <+75>:    mov    0xc(%rsp),%edi
   0x00007f926fde47cf <+79>:    mov    $0xe8,%eax
   0x00007f926fde47d4 <+84>:    syscall 
=> 0x00007f926fde47d6 <+86>:    cmp    $0xfffffffffffff000,%rax
   0x00007f926fde47dc <+92>:    ja     0x7f926fde4810 <epoll_wait+144>
   0x00007f926fde47de <+94>:    mov    %r8d,%edi
   0x00007f926fde47e1 <+97>:    mov    %eax,0xc(%rsp)
   0x00007f926fde47e5 <+101>:   call   0x7f926fd60dd0 <__libc_disable_asynccancel>
   0x00007f926fde47ea <+106>:   mov    0xc(%rsp),%eax
   0x00007f926fde47ee <+110>:   add    $0x28,%rsp
   0x00007f926fde47f2 <+114>:   ret    
   0x00007f926fde47f3 <+115>:   nopl   0x0(%rax,%rax,1)
   0x00007f926fde47f8 <+120>:   mov    0xc8649(%rip),%rdx        # 0x7f926feace48
   0x00007f926fde47ff <+127>:   neg    %eax
   0x00007f926fde4801 <+129>:   mov    %eax,%fs:(%rdx)
   0x00007f926fde4804 <+132>:   mov    $0xffffffff,%eax
   0x00007f926fde4809 <+137>:   ret    
   0x00007f926fde480a <+138>:   nopw   0x0(%rax,%rax,1)
   0x00007f926fde4810 <+144>:   mov    0xc8631(%rip),%rdx        # 0x7f926feace48
   0x00007f926fde4817 <+151>:   neg    %eax
   0x00007f926fde4819 <+153>:   mov    %eax,%fs:(%rdx)
   0x00007f926fde481c <+156>:   mov    $0xffffffff,%eax
   0x00007f926fde4821 <+161>:   jmp    0x7f926fde47de <epoll_wait+94>
End of assembler dump.
```

```
(gdb) i r
rax            0xfffffffffffffffc  -4
rbx            0xfffffffffffdfd90  -131696
rcx            0x7f926fde47d6      140266918791126
rdx            0xc8                200
rsi            0xc812d70           209792368
rdi            0x4                 4
rbp            0x2                 0x2
rsp            0x7ffe733b89c0      0x7ffe733b89c0
r8             0x0                 0
r9             0x1f                31
r10            0x2                 2
r11            0x293               659
r12            0xfffffffffffdfd90  -131696
r13            0x0                 0
r14            0xfffffffffffdd730  -141520
r15            0x2                 2
rip            0x7f926fde47d6      0x7f926fde47d6 <epoll_wait+86>
eflags         0x293               [ CF AF SF IF ]
cs             0x33                51
ss             0x2b                43
ds             0x0                 0
es             0x0                 0
fs             0x0                 0
gs             0x0                 0
k0             0x0                 0
k1             0x0                 0
k2             0x0                 0
k3             0x0                 0
k4             0x0                 0
k5             0x0                 0
k6             0x0                 0
k7             0x0                 0
```

**Comment by wtarreau** [comment]
Thanks, that's super strange! I must confess I have no idea. Also what is odd here is that it crashed on the return of the syscall so the reported instruction is not this one (cmp rax,-4096 cannot crash). Maybe it reported the wrong thread ? Just trying to guess :-/

**Comment by wtarreau** [comment]
Oh and yes please a separate issue would indeed help :-)  Don't waste your time on this one before the week-end, we know too well how addictive such issues can be.
